design/0005-circuit-breaking.md

```diff
@@ -0,0 +1,231 @@
+# Title
+
+- Contribution Name: circuit-breaking
+- Implementation Owner: @Ashish-Bansal, Linkerd Team
+- Start Date: 2020-06-14
+- Target Date: N.A.
+- Linkerd Issue:
+  [linkerd/linkerd2#2846](https://github.com/linkerd/linkerd2/issues/2846)
+- Reviewers: @adleong
+
+## Summary
+
+[summary]: #summary
+
+We would like to implement circuit breaking functionality which would allow end
+users to configure mesh such that they can limit the impact of failures from
+faulty service over others and prevent cascading issues. By configuring circuit
+breaking, they will be able to remove failing service instances from load
+balancing and survive through it in a more graceful manner.
+
+## Problem Statement
+
+In a service mesh, services often make calls to other services. In case upstream
+service is unable to respond due to any reason, it's possible that these
+failures can cascade to other services in the cluster.
+
+If we are aware that specific service instance is not responding, we can call it
+unhealthy and it's better if we just stop sending "traffic"[0] to that instance
+rather than getting just another expected failure.
+
+---
+
+It would also be somewhat helpful in case of canary deployments when there's
+some regression and new build is raising failures.
+
+[Example](https://drive.google.com/open?id=1FxKi-egyhHumCvlHvTI4P114z7qOLPeN)
+
+---
+
+Users should also be able to set max-ejection-percentage which provides max
+number of service instances which can be removed from the load balancing for
+specific service.
+
+[Example](https://drive.google.com/open?id=1jSMaogdvzxjv0v2dBoL7OYGHmsCusXn-)
+
+---
+
+Network Parition - There can be lot of scenarios in network partition but to
+provide a glimpse how circuit breaking would behave in case of network issue -
+[Example](https://drive.google.com/open?id=154I2ZGlGc-rQJ0YVeek17dJy7raAvd-G)
+
+---
+
+Protocol Specific features -
+
+1. For HTTP/gRPC based communication, users should be able to configure circuit
+   breaking either per endpoint or whole service.
+   [Example](https://drive.google.com/open?id=1JK65u9Mnmkmy_X8fczeh6RO6O-vNiTtY)
+   User can configure ejection based on many different metrics. These metrics
+   would be calculated over a given time range.
+   - Success Rate
+   - Percentile Latencies
+   - Max Requests
+   - Consecutive Failures
+2. For TCP based communication -
+   [Example](https://drive.google.com/open?id=1yAAurygsZQ-r58TEA0QiFf9Sz15jqRJV)
+
+---
```

Comment:
> Once the circuit is broken, what should be returned to the client? In all these cases a client should get some kind of notification.

Response sent to the client depends on kind of request. In case of HTTP, it would be 503 with some custom header in the response. For gRPC, it will be some kind of gRPC message.

Reply:
ToDo - add `ejectAllPercentageMark` -

Eject all instances once x% of upstream are already ejected. 

Usecase explanation -- it will be helpful when user know remaining % of upstreams won't be able to handle the incoming traffic and they are going to eventually fail.